07_web_endpoints/discord_bot.py

```diff
@@ -0,0 +1,230 @@
+# Discord within the time limit.
+
+
+@app.function(image=image)
+async def get_weather_forecast_for_city(
+    city: str, interaction_token, application_id
+):
+    import aiohttp
+    from python_weather import IMPERIAL, Client, Error, RequestError
+
+    interaction_url = f"https://discord.com/api/v10/webhooks/{application_id}/{interaction_token}/messages/@original"
+    async with Client(unit=IMPERIAL) as client:
+        try:
+            weather = await client.get(city)
+            daily_forecasts = "\n".join(
+                [
+                    f"Date: {daily.date}, Highest temperature: {daily.highest_temperature}°F, Lowest Temperature: {daily.lowest_temperature}°F"
+                    for daily in weather
+                ]
+            )
+            message = f"The forecast for {weather.location} is as follows:\n{daily_forecasts}"
+        except RequestError:
+            message = "An error occurred, issue with connecting to weather api"
+        except Error:
+            message = "An error occurred, please check city name"
+
+    json_payload = {"content": message}
+    async with aiohttp.ClientSession() as session:
+        async with session.patch(interaction_url, json=json_payload) as resp:
+            print(await resp.text())
+
+
+# We now define an asgi app using the Modal ASGI syntax [@asgi_app](https://modal.com/docs/guide/webhooks#serving-asgi-and-wsgi-apps).
+
+
+@app.function(
+    secrets=[modal.Secret.from_name("weather-discord-secret")],
+    keep_warm=1,  # eliminates risk of container startup making discord ack time too long
+    image=image,
+)
+@modal.asgi_app()
+def web_app():
+    from fastapi import FastAPI, HTTPException, Request
+    from fastapi.middleware.cors import CORSMiddleware
+
+    web_app = FastAPI()
+
+    # allow CORS
+    web_app.add_middleware(
+        CORSMiddleware,
+        allow_origins=["*"],
+        allow_credentials=True,
+        allow_methods=["*"],
+        allow_headers=["*"],
+    )
+
+    @web_app.post("/get_weather_forecast")
+    async def get_weather_forecast_api(request: Request):
+        import os
+
+        from nacl.exceptions import BadSignatureError
+        from nacl.signing import VerifyKey
+
+        # Verify the request using the Discord public key
+        public_key = os.getenv("DISCORD_PUBLIC_KEY")
+        verify_key = VerifyKey(bytes.fromhex(public_key))
+
+        signature = request.headers.get("X-Signature-Ed25519")
+        timestamp = request.headers.get("X-Signature-Timestamp")
+        body = await request.body()
+
+        message = timestamp.encode() + body
+
+        try:
+            verify_key.verify(message, bytes.fromhex(signature))
+        except BadSignatureError:
+            raise HTTPException(status_code=401, detail="Invalid request")
+
+        # Parse request
+        data = json.loads(body.decode())
+        if data.get("type") == 1:  # ack ping from Discord
```

Comment:
An enum would be good here, see [this code](https://github.com/the-full-stack/ask-fsdl/blob/b725c0ecdf0d484b48b79e32c37905614522ba4d/bot.py#L23-L34).